File: lint-arwaky/archive/capabilities/surface_hierarchy_checker.py
"""SurfaceHierarchyChecker — AES018/AES019 for surface hierarchy enforcement.

AES018 SURFACE_HIERARCHY_VIOLATION:
A file that is NOT an __init__.py barrel in the surfaces layer is not
imported from the layer __init__.py — meaning it is unreachable from the
surface entry point.

AES019 PASSIVE_SURFACE_VIOLATION:
A surface file contains complex domain logic (many public methods, deep
control flow) instead of acting as a thin pass-through to the agent layer.
Surfaces must be declarative/passive — I/O parsing + delegation only.
"""
from __future__ import annotations

import ast

from ..taxonomy import (
    ColumnNumber,
    ErrorCode,
    FilePath,
    FilePathList,
    LintMessage,
    LintResult,
    LintResultList,
    LineNumber,
    Severity,
)


class SurfaceHierarchyChecker:
    """AES018 + AES019 — surface barrel wiring and passivity checks."""

    # Thresholds for AES019
    _MAX_PUBLIC_METHODS = 10
    _MAX_FUNCTION_BODY_LINES = 80
    _MAX_IF_DEPTH = 3

    def check_surface_hierarchy(
        self,
        analyzer,
        files: FilePathList,
        root_dir: FilePath,
        results: LintResultList,
    ) -> None:
        """Main entry point — run AES018 (barrel wiring) and AES019 (passive surface)."""
        for f in files:
            if not _is_in_surfaces(f):
                continue
            if _is_init(f):
                continue

            basename = str(analyzer.fs.get_basename(f))
            layer_vo = analyzer._detect_layer(f, root_dir)
            if layer_vo:
                definition = analyzer.layer_map.get(layer_vo)
                if definition and definition.exceptions.values and basename in definition.exceptions.values:
                    continue

            # AES018: check if file is wired in barrel
            if not _is_wired(f):
                desc = (
                    f"AES018 SURFACE_HIERARCHY_VIOLATION: Surface file "
                    f"'{f.value}' is not imported from the layer barrel.\n"
                    f"WHY? All surface files must be reachable through __init__.py "
                    f"to maintain a clear entry-point hierarchy.\n"
                    f"FIX: Add 'from .{_stem(f)} import ...' to "
                    f"{_directory(f)}/__init__.py, or delete if unused."
                )
                results.values.append(
                    LintResult(
                        code=ErrorCode(code="AES018"),
                        message=LintMessage(value=desc),
                        severity=Severity.CRITICAL,
                        file=f,
                        line=LineNumber(value=1),
                        column=ColumnNumber(value=1),
                    )
                )

            # AES019: check if file is passive
            self._check_passive(f, results)

    def _check_passive(self, f: FilePath, results: LintResultList) -> None:
        """Check if a surface file is passive (thin I/O boundary)."""
        try:
            content = open(f.value, "r", encoding="utf-8").read()
            tree = ast.parse(content, filename=f.value)
        except (SyntaxError, UnicodeDecodeError, OSError):
            return

        violations: list[str] = []
        for node in ast.walk(tree):
            if not isinstance(node, ast.ClassDef):
                continue
            pub_methods = _collect_public_methods(node)
            self._check_methods_too_public(node, pub_methods, violations)
            self._check_method_lengths(node, pub_methods, violations)
            self._check_method_nesting(node, pub_methods, violations)

        if violations:
            self._report_aes019(f, violations, results)

    # -- AES019 sub-checks (each isolated to reduce CC) -------------------------

    def _check_methods_too_public(
        self, node: ast.ClassDef, pub_methods: list[ast.FunctionDef | ast.AsyncFunctionDef], violations: list[str],
    ) -> None:
        """AES019: too many public methods in a surface class."""
        if len(pub_methods) > self._MAX_PUBLIC_METHODS:
            violations.append(
                f"Class '{node.name}' has {len(pub_methods)} public "
                f"methods (max {self._MAX_PUBLIC_METHODS})"
            )

    def _check_method_lengths(
        self, node: ast.ClassDef, pub_methods: list[ast.FunctionDef | ast.AsyncFunctionDef], violations: list[str],
    ) -> None:
        """AES019: method body exceeds line limit."""
        for method in pub_methods:
            body_len = (method.end_lineno or 0) - method.lineno
            if body_len > self._MAX_FUNCTION_BODY_LINES:
                violations.append(
                    f"Method '{node.name}.{method.name}' is {body_len} lines "
                    f"(max {self._MAX_FUNCTION_BODY_LINES})"
                )

    def _check_method_nesting(
        self, node: ast.ClassDef, pub_methods: list[ast.FunctionDef | ast.AsyncFunctionDef], violations: list[str],
    ) -> None:
        """AES019: method control-flow nesting exceeds limit."""
        for method in pub_methods:
            for child in ast.walk(method):
                if isinstance(child, ast.If) and _if_depth(child) > self._MAX_IF_DEPTH:
                    violations.append(
                        f"Method '{node.name}.{method.name}' has deep "
                        f"control flow (if-nesting > {self._MAX_IF_DEPTH})"
                    )
                    break

    def _report_aes019(
        self, f: FilePath, violations: list[str], results: LintResultList,
    ) -> None:
        """Append a single AES019 result to the results list."""
        detail = "\n".join(f"  - {v}" for v in violations)
        results.values.append(
            LintResult(
                code=ErrorCode(code="AES019"),
                message=LintMessage(
                    value=(
                        f"AES019 PASSIVE_SURFACE_VIOLATION: Surface file "
                        f"'{f.value}' contains active domain logic:\n{detail}\n"
                        f"WHY? Surfaces must be passive I/O boundaries. "
                        f"Business logic belongs in capabilities/agent layers.\n"
                        f"FIX: Move logic to a handler or orchestrator."
                    )
                ),
                severity=Severity.CRITICAL,
                file=f,
                line=LineNumber(value=1),
                column=ColumnNumber(value=1),
            )
        )


# --- helpers -----------------------------------------------------------------

def _collect_public_methods(node: ast.ClassDef) -> list[ast.FunctionDef | ast.AsyncFunctionDef]:
    """Return public (non-underscore) function/method definitions from a class body."""
    return [
        n
        for n in node.body
        if isinstance(n, (ast.FunctionDef, ast.AsyncFunctionDef))
        and not n.name.startswith("_")
    ]


def _is_in_surfaces(f: FilePath) -> bool:
    return "/surfaces/" in f.value or f.value.endswith("/surfaces")


def _is_init(f: FilePath) -> bool:
    return f.value.endswith("__init__.py") or f.value.endswith("mod.rs") or f.value.endswith("index.ts") or f.value.endswith("index.js")


def _stem(f: FilePath) -> str:
    return f.value.rsplit("/", 1)[-1].rsplit(".", 1)[0]


def _directory(f: FilePath) -> str:
    return f.value.rsplit("/", 1)[0]


def _is_wired(f: FilePath) -> bool:
    """Return True if module stem is imported in its directory barrel."""
    barrel_names = ["__init__.py", "mod.rs", "index.ts", "index.js"]
    for name in barrel_names:
        init_path = _directory(f) + "/" + name
        try:
            content = open(init_path, "r", encoding="utf-8").read()
            stem = _stem(f)
            if (
                f"import {stem}" in content
                or f"from .{stem}" in content
                or f'"{stem}"' in content
                or f"'{stem}'" in content
                or f"mod {stem}" in content
                or f"use {stem}" in content
            ):
                return True
        except (FileNotFoundError, OSError):
            continue
    return False


def _if_depth(node: ast.If, depth: int = 0) -> int:
    max_d = depth
    for child in ast.iter_child_nodes(node):
        if isinstance(child, ast.If):
            max_d = max(max_d, _if_depth(child, depth + 1))
    return max_d

File: lint-arwaky/src-rust/capabilities/surface_hierarchy_checker.rs
// surface_hierarchy_checker — AES018/AES019 for surface hierarchy enforcement.
//
// AES018 SURFACE_HIERARCHY_VIOLATION:
// A file that is NOT an __init__.py barrel in the surfaces layer is not
// imported from the layer __init__.py — meaning it is unreachable from the
// surface entry point.
//
// AES019 PASSIVE_SURFACE_VIOLATION:
// A surface file contains complex domain logic (many public methods, deep
// control flow) instead of acting as a thin pass-through to the agent layer.
// Surfaces must be declarative/passive — I/O parsing + delegation only.

use crate::taxonomy::*;
use once_cell::sync::Lazy;
use regex::Regex;

// Regex: detect Python function/method definitions inside a class
static PY_METHOD_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^(?:async\s+)?def\s+(\w+)\s*\(").unwrap()
});

// Regex: detect class definitions
static PY_CLASS_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^class\s+(\w+)").unwrap()
});

// Regex: detect if statements for nesting depth
static IF_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^\s*if\s+").unwrap()
});

/// AES018 + AES019 — surface barrel wiring and passivity checks.
pub struct SurfaceHierarchyChecker;

// Thresholds for AES019
const MAX_PUBLIC_METHODS: usize = 10;
const MAX_FUNCTION_BODY_LINES: i64 = 80;
const MAX_IF_DEPTH: usize = 3;

impl SurfaceHierarchyChecker {
    pub fn new() -> Self {
        Self
    }

    /// Main entry point — run AES018 (barrel wiring) and AES019 (passive surface).
    pub fn check_surface_hierarchy(
        &self,
        files: &[FilePath],
        root_dir: &FilePath,
        results: &mut LintResultList,
    ) {
        for f in files {
            if !is_in_surfaces(f) {
                continue;
            }
            if is_init(f) {
                continue;
            }

            // AES018: check if file is wired in barrel
            if !is_wired(f) {
                let desc = format!(
                    "AES018 SURFACE_HIERARCHY_VIOLATION: Surface file '{}' is not imported from the layer barrel.\n\
                     WHY? All surface files must be reachable through __init__.py to maintain a clear entry-point hierarchy.\n\
                     FIX: Add 'from .{} import ...' to {}/__init__.py, or delete if unused.",
                    f,
                    stem(f),
                    directory(f)
                );
                results.append(LintResult {
                    file: f.clone(),
                    line: LineNumber::new(1),
                    column: ColumnNumber::new(1),
                    code: ErrorCode::new("AES018").unwrap(),
                    message: LintMessage::new(desc),
                    source: Some(AdapterName::new("surface_hierarchy").unwrap()),
                    severity: Severity::CRITICAL,
                    enclosing_scope: None,
                    related_locations: LocationList::new(),
                });
            }

            // AES019: check if file is passive
            self._check_passive(f, results);
        }
    }

    /// Check if a surface file is passive (thin I/O boundary).
    fn _check_passive(&self, f: &FilePath, results: &mut LintResultList) {
        let content = match std::fs::read_to_string(f.to_string()) {
            Ok(c) => c,
            Err(_) => return,
        };

        let lines: Vec<&str> = content.lines().collect();
        let mut violations: Vec<String> = Vec::new();

        // Find classes in the file and check their methods
        for (i, raw_line) in lines.iter().enumerate() {
            let stripped = raw_line.trim();
            if let Some(cap) = PY_CLASS_RE.captures(stripped) {
                let class_name = cap.get(1).map(|m| m.as_str()).unwrap_or("");
                let class_start = i;
                let indent = raw_line.len() - raw_line.trim_start().len();

                // Collect public methods in this class
                let mut pub_methods: Vec<(String, usize, Option<usize>)> = Vec::new();

                for j in (i + 1)..lines.len() {
                    let method_line = lines[j];
                    if method_line.trim().is_empty() {
                        continue;
                    }
                    let m_indent = method_line.len() - method_line.trim_start().len();

                    // If indent <= class indent, we've left the class
                    if m_indent <= indent && !method_line.trim().is_empty() {
                        break;
                    }

                    if let Some(mcap) = PY_METHOD_RE.captures(method_line.trim()) {
                        let method_name = mcap.get(1).map(|m| m.as_str()).unwrap_or("");
                        // Public methods don't start with underscore
                        if !method_name.starts_with('_') {
                            // Estimate end line
                            let mut end_line = lines.len();
                            for k in (j + 1)..lines.len() {
                                let next = lines[k];
                                if !next.trim().is_empty() {
                                    let n_indent = next.len() - next.trim_start().len();
                                    if n_indent <= m_indent {
                                        end_line = k;
                                        break;
                                    }
                                }
                            }
                            pub_methods.push((method_name.to_string(), j + 1, Some(end_line)));
                        }
                    }
                }

                self._check_methods_too_public(class_name, &pub_methods, &mut violations);
                self._check_method_lengths(class_name, &lines, &pub_methods, &mut violations);
                self._check_method_nesting(class_name, &lines, &pub_methods, &mut violations);
            }
        }

        if !violations.is_empty() {
            self._report_aes019(f, violations, results);
        }
    }

    // -- AES019 sub-checks ---------------------------------------------------

    /// AES019: too many public methods in a surface class.
    fn _check_methods_too_public(
        &self,
        class_name: &str,
        pub_methods: &[(String, usize, Option<usize>)],
        violations: &mut Vec<String>,
    ) {
        if pub_methods.len() > MAX_PUBLIC_METHODS {
            violations.push(format!(
                "Class '{}' has {} public methods (max {})",
                class_name,
                pub_methods.len(),
                MAX_PUBLIC_METHODS
            ));
        }
    }

    /// AES019: method body exceeds line limit.
    fn _check_method_lengths(
        &self,
        class_name: &str,
        _lines: &[&str],
        pub_methods: &[(String, usize, Option<usize>)],
        violations: &mut Vec<String>,
    ) {
        for (method_name, start, end) in pub_methods {
            if let Some(end_line) = end {
                let body_len = (*end_line as i64) - (*start as i64);
                if body_len > MAX_FUNCTION_BODY_LINES {
                    violations.push(format!(
                        "Method '{}.{}' is {} lines (max {})",
                        class_name, method_name, body_len, MAX_FUNCTION_BODY_LINES
                    ));
                }
            }
        }
    }

    /// AES019: method control-flow nesting exceeds limit.
    fn _check_method_nesting(
        &self,
        class_name: &str,
        lines: &[&str],
        pub_methods: &[(String, usize, Option<usize>)],
        violations: &mut Vec<String>,
    ) {
        for (method_name, start, end) in pub_methods {
            let end_line = end.unwrap_or(lines.len());
            let mut max_depth: usize = 0;

            for i in *start..end_line {
                if i >= lines.len() {
                    break;
                }
                let line = lines[i];
                let trimmed = line.trim();

                // Count nesting by indentation increase relative to method body
                if IF_RE.is_match(trimmed) {
                    let indent = line.len() - line.trim_start().len();
                    // Simple heuristic: count leading whitespace / 4 as depth
                    let depth = indent / 4;
                    if depth > max_depth {
                        max_depth = depth;
                    }
                }
            }

            if max_depth > MAX_IF_DEPTH {
                violations.push(format!(
                    "Method '{}.{}' has deep control flow (if-nesting > {})",
                    class_name, method_name, MAX_IF_DEPTH
                ));
            }
        }
    }

    /// Append a single AES019 result to the results list.
    fn _report_aes019(
        &self,
        f: &FilePath,
        violations: Vec<String>,
        results: &mut LintResultList,
    ) {
        let detail: String = violations
            .iter()
            .map(|v| format!("  - {}", v))
            .collect::<Vec<_>>()
            .join("\n");

        results.append(LintResult {
            file: f.clone(),
            line: LineNumber::new(1),
            column: ColumnNumber::new(1),
            code: ErrorCode::new("AES019").unwrap(),
            message: LintMessage::new(format!(
                "AES019 PASSIVE_SURFACE_VIOLATION: Surface file '{}' contains active domain logic:\n{}\n\
                 WHY? Surfaces must be passive I/O boundaries. Business logic belongs in capabilities/agent layers.\n\
                 FIX: Move logic to a handler or orchestrator.",
                f, detail
            )),
            source: Some(AdapterName::new("surface_hierarchy").unwrap()),
            severity: Severity::CRITICAL,
            enclosing_scope: None,
            related_locations: LocationList::new(),
        });
    }
}

// --- helpers -----------------------------------------------------------------

/// Check if the file path is in a surfaces directory.
fn is_in_surfaces(f: &FilePath) -> bool {
    f.to_string().contains("/surfaces/") || f.to_string().ends_with("/surfaces")
}

/// Check if the file is a barrel/init file.
fn is_init(f: &FilePath) -> bool {
    let path_str = f.to_string();
    path_str.ends_with("__init__.py")
        || path_str.ends_with("mod.rs")
        || path_str.ends_with("index.ts")
        || path_str.ends_with("index.js")
}

/// Extract the file stem (filename without extension).
fn stem(f: &FilePath) -> &str {
    let path_str = f.to_string();
    let basename = path_str.rsplit('/').next().unwrap_or(path_str);
    basename.rsplit('.').next().unwrap_or(basename)
}

/// Get the directory portion of the file path.
fn directory(f: &FilePath) -> &str {
    let path_str = f.to_string();
    path_str.rsplit('/').skip(1).next().unwrap_or(path_str)
}

/// Check if a module stem is imported in its directory barrel.
fn is_wired(f: &FilePath) -> bool {
    let barrel_names = ["__init__.py", "mod.rs", "index.ts", "index.js"];
    let file_stem = stem(f);
    let dir = directory(f);

    for name in &barrel_names {
        let init_path = format!("{}/{}", dir, name);
        if let Ok(content) = std::fs::read_to_string(&init_path) {
            if content.contains(&format!("import {}", file_stem))
                || content.contains(&format!("from .{}", file_stem))
                || content.contains(&format!("\"{}\"", file_stem))
                || content.contains(&format!("'{}'", file_stem))
                || content.contains(&format!("mod {}", file_stem))
                || content.contains(&format!("use {}", file_stem))
            {
                return true;
            }
        }
    }
    false
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_in_surfaces() {
        let f = FilePath::new("src/surfaces/handler.py").unwrap();
        assert!(is_in_surfaces(&f));

        let f = FilePath::new("src/capabilities/checker.py").unwrap();
        assert!(!is_in_surfaces(&f));
    }

    #[test]
    fn test_is_init() {
        let f = FilePath::new("src/surfaces/__init__.py").unwrap();
        assert!(is_init(&f));

        let f = FilePath::new("src/surfaces/handler.py").unwrap();
        assert!(!is_init(&f));
    }

    #[test]
    fn test_stem() {
        let f = FilePath::new("src/surfaces/handler.py").unwrap();
        assert_eq!(stem(&f), "handler");
    }

    #[test]
    fn test_directory() {
        let f = FilePath::new("src/surfaces/handler.py").unwrap();
        assert_eq!(directory(&f), "src/surfaces");
    }
}
